charms/kfp-ui/tests/unit/test_operator.py

```diff
@@ -12,327 +14,590 @@
-                "port": 1234,
-                "secret-key": "secret-key",
-                "secure": True,
-                "service": "service",
-            },
-            does_not_raise(),
-            None,
-        ),
-        # kfp-api
-        (
-            # No relation established.  Raises CheckFailedError
-            "kfp-api",
-            None,
-            None,
-            pytest.raises(CheckFailedError),
-            BlockedStatus("Missing required relation for kfp-api"),
-        ),
-        (
-            # Relation exists but no versions set yet
-            "kfp-api",
-            {},
-            None,
-            pytest.raises(CheckFailedError),
-            WaitingStatus(
-                "List of <ops.model.Relation kfp-api:0> versions not found for apps: other-app"
-            ),
-        ),
-        (
-            # Relation exists with versions, but no data posted yet
-            "kfp-api",
-            {"_supported_versions": "- v1"},
-            None,
-            pytest.raises(CheckFailedError),
-            WaitingStatus("Waiting for kfp-api relation data"),
-        ),
-        (
-            # Relation exists with versions and empty data
-            "kfp-api",
-            {"_supported_versions": "- v1", "data": yaml.dump({})},
-            None,
-            pytest.raises(CheckFailedError),
-            WaitingStatus("Waiting for kfp-api relation data"),
-        ),
-        (
-            # Relation exists with versions and invalid (partial) data
-            "kfp-api",
-            {
-                "_supported_versions": "- v1",
-                "data": yaml.dump({"service-name": "service-name"}),
-            },
-            None,
-            pytest.raises(CheckFailedError),
-            BlockedStatus("Failed to validate data on kfp-api:0 from other-app"),
-        ),
-        (
-            # Relation exists with valid data
-            "kfp-api",
-            {
-                "_supported_versions": "- v1",
-                "data": yaml.dump(
-                    {
-                        "service-name": "service-name",
-                        "service-port": "1234",
-                    }
-                ),
-            },
-            {
-                "service-name": "service-name",
-                "service-port": "1234",
-            },
-            does_not_raise(),
-            None,
-        ),
-    ),
-)
-def test_relations_that_provide_data(
-    harness,
-    relation_name,
-    relation_data,
-    expected_returned_data,
-    expected_raises,
-    expected_status,
-):
-    harness.set_leader()
+
+    # Need to mock the leadership-gate to be active, and the kubernetes auth component so that it
+    # sees the expected resources when calling _get_missing_kubernetes_resources
+
+    harness.charm.leadership_gate_component_item.executed = True
+    harness.charm.leadership_gate_component_item.get_status = MagicMock(return_value=ActiveStatus())
```

Comment:
This looks redundant since you already do `harness.set_leader(True)`

Reply:
yeah it really is.  I wanted to avoid mocking parts of the charm and not others, but you're completely right this is redundant.